pgjdbc/src/main/java/org/postgresql/gss/GSSOutputStream.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2020, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.gss;
+
+import org.ietf.jgss.GSSContext;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.MessageProp;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+public class GSSOutputStream extends OutputStream {
+  private final GSSContext gssContext;
+  private final MessageProp messageProp;
+  private byte[] buffer;
+  private byte[] int4Buf = new byte[4];
+  private int index;
+  private OutputStream wrapped;
+
+  public GSSOutputStream(OutputStream out, GSSContext gssContext, MessageProp messageProp, int bufferSize)  {
+    wrapped = out;
+    this.gssContext = gssContext;
+    this.messageProp = messageProp;
+    buffer = new byte[bufferSize];
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    buffer[index++] = (byte)b;
+    if (index >= buffer.length) {
+      flush();
+    }
+  }
+
+  @Override
+  public void write(byte[] buf) throws IOException {
+    write(buf, 0, buf.length);
+  }
+
+  @Override
+  public void write(byte[] b, int pos, int len) throws IOException {
+    int max;
+
+    while ( len > 0 ) {
+      int roomToWrite = buffer.length - index;
+      if ( len < roomToWrite ) {
+        System.arraycopy(b, pos,buffer, index, len);
+        index += len;
+        len -= roomToWrite;
+      } else {
+        System.arraycopy(b, pos, buffer, index, roomToWrite );
+        index += roomToWrite;
+        len -= roomToWrite;
+      }
+      if (roomToWrite == 0) {
+        flush();
```

Comment:
https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-GSSAPI

> Note that the server will only accept encrypted packets from the client which are less than 16kB; gss_wrap_size_limit() should be used by the client to determine the size of the unencrypted message which will fit within this limit and larger messages should be broken up into multiple gss_wrap() calls

It looks like we need to use `getWrapSizeLimit` to determine the max input raw message length to fit within 16KiB of encrypted data.

@davecramer , do you remember why do you chunk the writes by the buffer boundaries?
What if the user requests to write 1MiB? Would it be better to write full buffer is a single go rather than chunk it into 16KiB pieces?

Reply:
I'm surprised this worked if that's the case ?

Reply:
Here's the backend enforcing max message length: https://github.com/postgres/postgres/blob/acecd6746cdc2df5ba8dcc2c2307c6560c7c2492/src/backend/libpq/be-secure-gssapi.c#L348

It ensures the GGS message length does not exceed `PQ_GSS_RECV_BUFFER_SIZE` (`16384`), and the `int4 messageSize` counts towards the limit.

I guess we just did not have tests where we sent large amount of data with GSS connection.

Reply:
nor did anyone else :( as nobody has complained about it not working.

Reply:
I came across this code as I was trying to unify buffers in https://github.com/pgjdbc/pgjdbc/pull/3248. I want to have a common superclass so we can have methods like `writeInt4`.
I hope PR 3248 improves things, however, testing GSS seems non-trivial.

Reply:
It most certainly is. I used to be able to test it locally. Put your changes in and I'll work on testing it again.

Reply:
https://github.com/pgjdbc/pgjdbc/pull/3248 contains all the relevant changes.
I wonder if you could somehow execute all pgjdbc tests and force it run everything through GSS.

Reply:
I won't get to this today. That is an interesting idea.
I'm close to being able to provide a PR to enable ALPN SSL connections.